connector/src/rest/connector/libs/dcnm/implementation.py

```diff
@@ -0,0 +1,370 @@
+        self.session.auth = HTTPBasicAuth(username, password)
+        response = self.session.post(login_url, data=json.dumps({'expirationTime': 999999}), headers = STD_HEADER, verify=self.verify)
+        log.info(response)
+
+        # Make sure it returned requests.codes.ok
+        if response.status_code != requests.codes.ok:
+            # Something bad happened
+            raise RequestException("Connection to '{ip}' has returned the "
+                                   "following code '{c}', instead of the "
+                                   "expected status code '{ok}'" \
+                                   .format(ip=host, c=response.status_code,
+                                           ok=requests.codes.ok))
+
+        self.token = response.json()['Dcnm-Token']  # Retrieve the Token from the returned JSONhahhah
+
+        log.info("Connected successfully to '{d}'".format(d=self.device.name))
+        return self.token
+
+
+    @BaseConnection.locked
+    def disconnect(self):
+        '''disconnect the device for this particular alias'''
+
+        log.info("Disconnecting from '{d}' with "
+                 "alias '{a}'".format(d=self.device.name, a=self.alias))
+        try:
+            self.session.close()
+        finally:
+            self._is_connected = False
+        log.info("Disconnected successfully from "
+                 "'{d}'".format(d=self.device.name))
+
+    def isconnected(func):
+        '''Decorator to make sure session to device is active
+
+           There is limitation on the amount of time the session ca be active
+           for on the DCNM devices. However, there are no way to verify if
+           session is still active unless sending a command. So, its just
+           faster to reconnect every time.
+         '''
+        def decorated(self, *args, **kwargs):
+            # Check if connected
+            try:
+                log.propagate = False
+                self.disconnect()
+                self.connect()
+                log.propagate = True
+                print(kwargs)
+                ret = func(self, *args, **kwargs)
+            finally:
+                log.propagate = True
+            return ret
+        return decorated
+
+    @BaseConnection.locked
+    def _request(self, method, api_url, **kwargs):
+        """ Wrapper to send REST command to device
+
+        Args:
+            method (str): session request method
+
+            dn (str): rest endpoint
+
+        Returns:
+            response.json() or response.text
+
+        Raises:
+            RequestException if response is not ok
+
+        """
+        if not self.connected:
+            raise Exception("'{d}' is not connected for alias '{a}'"
+                            .format(d=self.device.name,
+                                    a=self.alias))
+
+        # Deal with the dn
+        full_url = '{f}{api_url}'.format(f=self.url, api_url=api_url)
+
+        if 'data' in kwargs:
+            p = kwargs['data']
```

Comment:
instead of p as a variable name, can we name it payload ? Easier to read

Reply:
Addressed